Sources/performance-worker/BenchmarkService.swift

```diff
@@ -0,0 +1,185 @@
+  ) async throws
+    -> ServerResponse.Single<Grpc_Testing_BenchmarkService.Method.StreamingFromClient.Output>
+  {
+    var responseSize = 0
+    for try await message in request.messages {
+      if message.responseStatus.isInitialized {
+        try self.checkOkStatus(message.responseStatus)
+      }
+      responseSize = Int(message.responseSize)
+    }
+
+    return ServerResponse.Single(
+      message: Grpc_Testing_BenchmarkService.Method.StreamingFromClient.Output.with {
+        $0.payload = Grpc_Testing_Payload.with {
+          $0.body = Data(count: responseSize)
+        }
+      }
+    )
+  }
+
+  /// Single-sided unbounded streaming from server to client.
+  /// The server repeatedly returns a payload with the size requested by the client.
+  func streamingFromServer(
+    request: ServerRequest.Single<Grpc_Testing_BenchmarkService.Method.StreamingFromServer.Input>
+  ) async throws
+    -> ServerResponse.Stream<Grpc_Testing_BenchmarkService.Method.StreamingFromServer.Output>
+  {
+    if request.message.responseStatus.isInitialized {
+      try self.checkOkStatus(request.message.responseStatus)
+    }
+    let response = Grpc_Testing_BenchmarkService.Method.StreamingCall.Output.with {
+      $0.payload = Grpc_Testing_Payload.with {
+        $0.body = Data(count: Int(request.message.responseSize))
+      }
+    }
+    return ServerResponse.Stream { writer in
+      while working.load(ordering: .acquiring) {
+        try await writer.write(response)
+      }
+      return [:]
+    }
+  }
+
+  /// Two-sided unbounded streaming between server to client.
+  /// Both sides send the content of their own choice to the other.
+  func streamingBothWays(
+    request: GRPCCore.ServerRequest.Stream<
+      Grpc_Testing_BenchmarkService.Method.StreamingBothWays.Input
+    >
+  ) async throws
+    -> ServerResponse.Stream<Grpc_Testing_BenchmarkService.Method.StreamingBothWays.Output>
+  {
+    // The 100 size is used by the other implementations as well.
+    // We are using the same canned response size for all responses
+    // as it is allowed by the spec.
+    let response = Grpc_Testing_BenchmarkService.Method.StreamingCall.Output.with {
+      $0.payload = Grpc_Testing_Payload.with {
+        $0.body = Data(count: 100)
+      }
+    }
+
+    var serverResponse = ServerResponse.Stream<
+      Grpc_Testing_BenchmarkService.Method.StreamingBothWays.Output
+    >(accepted: .failure(RPCError(code: .internalError, message: "There was a server issue.")))
+
+    try await withThrowingTaskGroup(
+      of: ServerResponse.Stream<Grpc_Testing_BenchmarkService.Method.StreamingBothWays.Output>?.self
+    ) { group in
+      group.addTask {
+        for try await message in request.messages {
+          if message.responseStatus.isInitialized {
+            try self.checkOkStatus(message.responseStatus)
+          }
+        }
+        _ = self.working.exchange(false, ordering: .acquiring)
```

Comment:
I don't think this is quite right: `self.working` will be toggled by the quit method on the worker service, it's an external trigger to shutdown. Looking at it in a different way: I don't think it makes sense that the end of the request stream for this method would stop the response stream from streaming-from-server. 

Instead I think we want a separate atomic to signal between the request stream and response stream.

Reply:
Thank you! It makes sense to use a different atomic for this. One question though: I should still check when writing the responses if 'working' is set to 'true', besides checking the new atomic, right?